### PR TITLE
Replace most uses of Map by HashMap

### PR DESCRIPTION
**Pull request by reduz**
* Map is unnecessary and inefficient in almost every case.
* Replaced by the new HashMap
* Renamed Map to RBMap and RBSet for cases that still make sense (order matters) but use is discouraged.

There were very few cases where replacing by HashMap was undesired because keeping the key order was intended. I tried to keep those (as RBMap) as much as possible, but might have missed some. Review appreciated!

*Bugsquad edit:* Follow-up to #60881.

### CODE REVIEW

File: platform/iphone/tts_ios.h [hunk 1]
```diff
@@ -39,13 +39,13 @@
 
 #include "core/string/ustring.h"
 #include "core/templates/list.h"
-#include "core/templates/map.h"
+#include "core/templates/rb_map.h"
```
Comment (bruvzg) [inline platform/iphone/tts_ios.h:42]:
```suggestion
#include "core/templates/hash_map.h"
```

File: platform/linuxbsd/tts_linux.h [hunk 2]
```diff
@@ -35,7 +35,7 @@
 #include "core/os/thread_safe.h"
 #include "core/string/ustring.h"
 #include "core/templates/list.h"
-#include "core/templates/map.h"
+#include "core/templates/rb_map.h"
```
Comment (bruvzg) [inline platform/linuxbsd/tts_linux.h:38]: Same.

File: platform/windows/tts_windows.h [hunk 3]
```diff
@@ -33,7 +33,7 @@
 
 #include "core/string/ustring.h"
 #include "core/templates/list.h"
-#include "core/templates/map.h"
+#include "core/templates/rb_map.h"
```
Comment (bruvzg) [inline platform/windows/tts_windows.h:36]: Same.

File: scene/resources/font.h [hunk 4]
```diff
@@ -33,7 +33,7 @@
 
 #include "core/io/resource.h"
 #include "core/templates/lru.h"
-#include "core/templates/map.h"
+#include "core/templates/rb_map.h"
```
Comment (bruvzg) [inline scene/resources/font.h:36]: Probably not needed at all.

File: modules/text_server_adv/text_server_adv.h [hunk 5]
```diff
@@ -126,7 +126,7 @@ class TextServerAdvanced : public TextServerExtension {
 	_THREAD_SAFE_CLASS_
 
 	struct NumSystemData {
-		Set<StringName> lang;
+		RBSet<StringName> lang;
```
Comment (bruvzg) [inline modules/text_server_adv/text_server_adv.h:129]: Order is irrelevant for all `TextServer` sets.

**Review by bruvzg — COMMENTED** [review]
Seems like it's formatted with a wrong clang-format version, which is expanding single line functions.

**Comment by bruvzg** [comment]
Seems like you have missed `platform/osx/display_server_osx.mm` (and maybe other Objective-C files).

File: modules/webrtc/webrtc_multiplayer_peer.h [hunk 6]
```diff
@@ -69,7 +69,7 @@ class WebRTCMultiplayerPeer : public MultiplayerPeer {
 	int next_packet_peer = 0;
 	bool server_compat = false;
 
-	Map<int, Ref<ConnectedPeer>> peer_map;
+	HashMap<int, Ref<ConnectedPeer>> peer_map;
```
Comment (Faless) [inline modules/webrtc/webrtc_multiplayer_peer.h:72]: This is used to also balance incoming packets fairly among peers (in `_fin_next_peer`)
So I guess it shold be an `RBSet`/`RBMap`?

File: scene/gui/text_edit.h [hunk 7]
```diff
@@ -511,7 +511,7 @@ class TextEdit : public Control {
 
 	/* Syntax highlighting. */
 	Ref<SyntaxHighlighter> syntax_highlighter;
-	Map<int, Dictionary> syntax_highlighting_cache;
```
Comment (Paulb23) [inline scene/gui/text_edit.h:514]: `syntax_highlighting_cache` can be removed, as it's been moved to `SyntaxHighlighter`.

File: platform/windows/display_server_windows.cpp [hunk 8]
```diff
@@ -1501,7 +1501,7 @@ void DisplayServerWindows::cursor_set_custom_image(const Ref<Resource> &p_cursor
 	_THREAD_SAFE_METHOD_
 
 	if (p_cursor.is_valid()) {
-		Map<CursorShape, Vector<Variant>>::Element *cursor_c = cursors_cache.find(p_shape);
+		HashMap<CursorShape, Vector<Variant>>::Element *cursor_c = cursors_cache.find(p_shape);
 
 		if (cursor_c) {
 			if (cursor_c->get()[0] == p_cursor && cursor_c->get()[1] == p_hotspot) {
```
Comment (bruvzg) [inline platform/windows/display_server_windows.cpp:1507]:
```suggestion
		const HashMap<CursorShape, Vector<Variant>>::Iterator cursor_c = cursors_cache.find(p_shape);

		if (cursor_c) {
			if (cursor_c->value[0] == p_cursor && cursor_c->value[1] == p_hotspot) {
```

File: platform/android/tts_android.cpp [hunk 9]
```diff
@@ -175,7 +175,7 @@ void TTS_Android::resume() {
 }
 
 void TTS_Android::stop() {
-	for (Map<int, Char16String>::Element *E = ids.front(); E; E = E->next()) {
+	for (HashMap<int, Char16String>::Element *E = ids.front(); E; E = E->next()) {
 		DisplayServer::get_singleton()->tts_post_utterance_event(DisplayServer::TTS_UTTERANCE_CANCELED, E->key());
```
Comment (bruvzg) [inline platform/android/tts_android.cpp:179]:
```suggestion
	for (const KeyValue<int, Char16String> &E : ids) {
		DisplayServer::get_singleton()->tts_post_utterance_event(DisplayServer::TTS_UTTERANCE_CANCELED, E.key);
```

File: editor/import/dynamic_font_import_settings.h [hunk 10]
```diff
@@ -161,8 +161,8 @@ class DynamicFontImportSettings : public ConfirmationDialog {
 	Ref<Font> font_preview;
 	Ref<Font> font_main;
 
-	Set<char32_t> selected_chars;
-	Set<int32_t> selected_glyphs;
+	RBSet<char32_t> selected_chars;
+	RBSet<int32_t> selected_glyphs;
```
Comment (bruvzg) [inline editor/import/dynamic_font_import_settings.h:165]: Order is irrelevant.

File: modules/webrtc/webrtc_multiplayer_peer.h (hunk 6 above)
Comment (Faless) [reply modules/webrtc/webrtc_multiplayer_peer.h:72]: Not really I guess, it's more about keeping a consistent order.
That said though, given the number of peers doesn't change often (so the order won't change often), it's actually probably okay to stay as it is now.

**Comment by reduz** [comment]
This is now ready for review.

File: scene/gui/text_edit.h (hunk 7 above)
Comment (akien-mga) [reply scene/gui/text_edit.h:514]: Fixed.

File: editor/import/dynamic_font_import_settings.h (hunk 10 above)
Comment (akien-mga) [reply editor/import/dynamic_font_import_settings.h:165]: This PR only seems to rename `Set` to `RBSet`, it doesn't actually change its uses.

**Comment by akien-mga** [comment]
Amended to fix (manually..) the broken formatting caused by clang-format 14 in some headers.

File: editor/import/dynamic_font_import_settings.h (hunk 10 above)
Comment (bruvzg) [reply editor/import/dynamic_font_import_settings.h:165]: In any case, this part is rewritten for the `FontConfig`.

**Comment by akien-mga** [comment]
Did a quick sanity check as I had done for #60881. As mentioned there, it's not necessarily representative of what this PR impacts, it's just to make sure there's no huge obvious regression. The test is to start the editor on a 3D scene and then close it.

```
$ hyperfine -iw2 -m10 "bin/godot.linuxbsd.opt.tools.64.map_to_hashmap /home/akien/Projects/godot/projects/liblast/Game/project.godot" "bin/godot.linuxbsd.opt.tools.64.orig /home/akien/Projects/godot/projects/liblast/Game/project.godot"
Benchmark #1: bin/godot.linuxbsd.opt.tools.64.map_to_hashmap /home/akien/Projects/godot/projects/liblast/Game/project.godot
  Time (mean ± σ):     10.125 s ±  0.397 s    [User: 7.405 s, System: 0.479 s]
  Range (min … max):    9.458 s … 10.654 s    10 runs
 
Benchmark #2: bin/godot.linuxbsd.opt.tools.64.orig /home/akien/Projects/godot/projects/liblast/Game/project.godot
  Time (mean ± σ):      9.822 s ±  0.480 s    [User: 7.117 s, System: 0.491 s]
  Range (min … max):    9.016 s … 10.374 s    10 runs
 
Summary
  'bin/godot.linuxbsd.opt.tools.64.orig /home/akien/Projects/godot/projects/liblast/Game/project.godot' ran
    1.03 ± 0.06 times faster than 'bin/godot.linuxbsd.opt.tools.64.map_to_hashmap /home/akien/Projects/godot/projects/liblast/Game/project.godot'
```

The difference seems small enough to merge this for now and unblock other PRs.

There's likely still room for optimization and for doing proper benchmarks to see if/where there are bottlenecks. But generally speaking this PR mostly refactors the codebase to a better API so it can be merged as is, and the base HashMap / RBSet / RBMap implementations can be optimized further without changing the rest of the codebase.

**Review by bruvzg — APPROVED** [review]
Seems to run fine on macOS, and I do not see any obvious slowdowns.

_Note: There are a few questionable include changes (`map` → `rb_map` while code was changed to use `HashMap`)._

**Comment by akien-mga** [comment]
> Note: There are a few questionable include changes (map → rb_map while code was changed to use HashMap).

We might want to do a cleanup pass for `rb_map.h` and `rb_set.h` by comparing the outputs of:
```
rg RBMap -l | sort
rg rb_map.h -l | sort
```

**Comment by akien-mga** [comment]
Thanks!

**Comment by lufog** [comment]
@reduz, this PR broke `class_name`s.
[ClassNameTest_G4.zip](https://github.com/godotengine/godot/files/8702636/ClassNameTest_G4.zip) (in v4.0.alpha.custom_build [396def9b6] works fine)

**Review by Brayan-724 — COMMENTED** [review]
.

**Pull request merged**